pages/_app.tsx

```diff
@@ -1,20 +1,20 @@
 import CssBaseline from '@material-ui/core/CssBaseline'
 import { ThemeProvider, createMuiTheme } from '@material-ui/core/styles'
-import App from 'next/app'
 import Head from 'next/head'
-import React, { useState } from 'react'
+import React, { useState, useEffect } from 'react'
 import { red, blue, deepPurple } from '@material-ui/core/colors'
 import { Color } from '@material-ui/core'
-import { NextComponentType, NextPageContext } from 'next'
+import { NextPage, NextComponentType, NextPageContext } from 'next'
+import { AppProps } from 'next/app'
 
 import { AppContext } from '../components/AppContext'
 
 interface Props {
-	component: NextComponentType<NextPageContext, any, {}>
+	Component: NextComponentType<NextPageContext, any, {}>
 	pageProps: any
 }
 
-const AppWithHooks = (props: Props): JSX.Element => {
+const MyApp: NextPage<AppProps> = ({ Component, pageProps }: Props): JSX.Element => {
```

Comment:
This is better than `AppWithHooks` but I honestly think it would be best left as a class and just using context the classical way. I don't think themimg is enough of a justification to completely re-write this class as a function, especially when it [isn't necessary](https://www.taniarascia.com/using-context-api-in-react/#consuming-context).

Reply:
Classical components are considered legacy and have no support for fun things like `useEffect`. I think this rewrite only improves the codebase.

Reply:
> have no support for fun things like `useEffect`

That's deliberately misleading. All `useEffect` does is wrap the classical `componentDidMount`, `componentDidUpdate`, and `componentWillUnMount`. All `useState` does is wrap the classical `this.state`.

Reply:
`useEffect` contains a dependency array, which makes working with state in general much easier. It lets you listen for updates in the state for specific vars, which is much more painful in class components because you have to do all the checks in the `componentDidUpdate` function.